### PR TITLE
feature: support `native-image` for `seata-client`

### PR DESCRIPTION
**Pull request by wangliang181230**
- [ ] I have registered the PR [changes](https://github.com/seata/seata/tree/develop/changes).

### Ⅰ. Describe what this PR did
feature: First support `native-image` of the `seata-client`.
新特性：首次针对 `native-image` 做兼容性改造。

---

已支持的功能：
1. AT
2. XA
3. TCC
4. nacos: 注册中心nacos1.4.4测试通过，2.x测试未通过。配置中心暂未测试。
5. eureka: 使用seata中定义的版本时，测试通过。

暂未支持的功能：
1. SAGA
2. 其他可选组件，都未测试。

---

### Ⅳ. Describe how to verify it
1. 下载此PR的代码，并执行 `mvn clean install -DskipTests`，将 jar包 install 到本地仓库。
7. 下载我的学习用仓库的 `seata-support-aot` 分支：https://github.com/wangliang181230/study-spring-boot/tree/seata-support-aot
8. 安装打包native所需环境，参考教程：https://easyj.icu/blog/#/native-image/
9. 用 IDE 单独打开项目中的 `test-compatibility-springboot3-seata` 小目录，不是 `study-spring-boot` 大目录。
10. 运行命令 `mvn clean package -Pnative -e -DskipTests`，开始打包 `native-image`
   > 注意：打包 `native-image` 非常消耗内存，内存不足时，容易失败，请尽量关闭不需要的软件，释放内存。
   > 运行完成后，`/target/` 目录下，会生成一个 `exe` 文件，后面环境准备好后。
11. 启动 `seata-server`
12. 启动 `mysql`，并创建数据库：`seata_study`，再运行以下脚本创建表：
```sql
CREATE TABLE `tb_order` (
  `id` bigint NOT NULL,
  PRIMARY KEY (`id`)
) ENGINE=InnoDB;

CREATE TABLE `undo_log` (
  `branch_id` bigint NOT NULL COMMENT 'branch transaction id',
  `xid` varchar(128) NOT NULL COMMENT 'global transaction id',
  `context` varchar(128) NOT NULL COMMENT 'undo_log context,such as serialization',
  `rollback_info` longblob NOT NULL COMMENT 'rollback info',
  `log_status` int NOT NULL COMMENT '0:normal status,1:defense status',
  `log_created` datetime(6) NOT NULL COMMENT 'create datetime',
  `log_modified` datetime(6) NOT NULL COMMENT 'modify datetime',
  UNIQUE KEY `ux_undo_log` (`xid`,`branch_id`)
) ENGINE=InnoDB COMMENT='AT transaction mode undo table';
```
13. 点击目录下的 `startup-exe.bat` 文件，启动 `native-image`。
14. 访问站点根地址 `http://localhost:8081/`，测试成功情况；
15. 另外再访问  `http://localhost:8081/?test=1`，测试 insert 语句完成后，抛异常的情况，观察表 `undo_log` 中会有一条数据，等10秒后回滚后被删除。
16. 另外再访问  `http://localhost:8081/?test=2`，测试 insert 语句 和 `tcc.prepare` 方法完成后，抛异常的情况，观察表 `undo_log` 中会有一条数据，等10秒后回滚后被删除。同时，控制台会打印类似 `test tcc: rollback......` 字样。
17. 关闭 exe 进程，启动 `startup-exe-by-XA-mode.bat`，重复8、9、10步骤，测试 `XA` 模式。
   > 以上，AT、XA、TCC三种模式，测试成功。

### Ⅴ. Special notes for reviews
`seata-spring-framework-fake-for-java8` 模块，仅为了使aot相关代码在java8可以编译通过。使 `seata` 的jar包可以同时兼容高低版本的 `java`.

### CODE REVIEW

File: seata-spring-autoconfigure/seata-spring-autoconfigure-core/src/main/java/io/seata/spring/boot/autoconfigure/provider/SpringBootConfigurationProvider.java [hunk 1]
```diff
@@ -149,17 +148,31 @@ private Object getDefaultValueFromPropertyObject(String dataId) throws IllegalAc
...
+            Field field = propertyObj.getClass().getDeclaredField(fieldName);
 
-        // Get defaultValue from the field
-        if (fieldOptional.isPresent()) {
-            Field field = fieldOptional.get();
             if (!Map.class.isAssignableFrom(field.getType())) {
                 field.setAccessible(true);
                 return field.get(propertyObj);
             }
+        } catch (NoSuchFieldException e) {
```
Comment (wangliang181230) [inline seata-spring-autoconfigure/seata-spring-autoconfigure-core/src/main/java/io/seata/spring/boot/autoconfigure/provider/SpringBootConfigurationProvider.java:159]: 字段未找到时，就用方法获取。
因为在native-image中，有可能未注册fields，但注册了methods，可以通过method获取到配置值。

**Comment by wangliang181230** [comment]
@slievrly PTAL

**Review by slievrly — APPROVED** [review]
LGTM

**Review by xingfudeshi — APPROVED** [review]
LGTM

File: config/seata-config-core/src/main/java/io/seata/config/ConfigurationCache.java [hunk 2]
```diff
@@ -101,8 +100,8 @@ public void onChangeEvent(ConfigurationChangeEvent event) {
     }
 
     public Configuration proxy(Configuration originalConfiguration) throws Exception {
-        return new ByteBuddy().subclass(Configuration.class).method(ElementMatchers.any())
-            .intercept(InvocationHandlerAdapter.of((proxy, method, args) -> {
+        return (Configuration)Proxy.newProxyInstance(this.getClass().getClassLoader(), new Class[]{Configuration.class}
+            , (proxy, method, args) -> {
```
Comment (linghengqian) [inline config/seata-config-core/src/main/java/io/seata/config/ConfigurationCache.java:104]: - @funky-eyes
- As a solution to https://github.com/raphw/byte-buddy/issues/1588 , can we port the handling of `ConfigurationCache.java` back to the 2.x branch? 
- Reference https://github.com/apache/incubator-seata/blob/2.x/config/seata-config-core/src/main/java/org/apache/seata/config/ConfigurationCache.java .

Comment (funky-eyes) [reply config/seata-config-core/src/main/java/io/seata/config/ConfigurationCache.java:104]: > * @funky-eyes
> * As a solution to [No classes have been predefined during the image build to load from bytecodes at runtime raphw/byte-buddy#1588](https://github.com/raphw/byte-buddy/issues/1588) , can we port the handling of `ConfigurationCache.java` back to the 2.x branch?
> * Reference https://github.com/apache/incubator-seata/blob/2.x/config/seata-config-core/src/main/java/org/apache/seata/config/ConfigurationCache.java .

针对configuration cache代理这块可以修改为接口的方式使用jdk代理来解决，但是2.0中重构了事务api相关的模块，针对用户发起事务的实例使用了bytebuddy进行了代理，所以我不知道单纯改动configuration cache这个代理为jdk代理是否能解决该问题。
The configuration cache proxy can be modified to use a JDK proxy to address this issue. However, in 2.0, the transaction API-related modules were refactored, and ByteBuddy was used for proxying instances that initiate user transactions. Therefore, I am unsure whether simply changing the configuration cache proxy to a JDK proxy will resolve the issue.

Comment (funky-eyes) [reply config/seata-config-core/src/main/java/io/seata/config/ConfigurationCache.java:104]: > * @funky-eyes
> * As a solution to [No classes have been predefined during the image build to load from bytecodes at runtime raphw/byte-buddy#1588](https://github.com/raphw/byte-buddy/issues/1588) , can we port the handling of `ConfigurationCache.java` back to the 2.x branch?
> * Reference https://github.com/apache/incubator-seata/blob/2.x/config/seata-config-core/src/main/java/org/apache/seata/config/ConfigurationCache.java .

https://github.com/apache/incubator-seata/pull/6386  configurationcache will remove bytebuddy proxy in this pr

Comment (linghengqian) [reply config/seata-config-core/src/main/java/io/seata/config/ConfigurationCache.java:104]: > > * @funky-eyes
> > * As a solution to [No classes have been predefined during the image build to load from bytecodes at runtime raphw/byte-buddy#1588](https://github.com/raphw/byte-buddy/issues/1588) , can we port the handling of `ConfigurationCache.java` back to the 2.x branch?
> > * Reference https://github.com/apache/incubator-seata/blob/2.x/config/seata-config-core/src/main/java/org/apache/seata/config/ConfigurationCache.java .
> 
> #6386 configurationcache will remove bytebuddy proxy in this pr

- @funky-eyes  Late reply, I have verified that the relevant PR is enough to make the AT mode of Seata Client 2.1.0-SNAPSHOT work under GraalVM Native Image. I opened https://github.com/apache/incubator-seata/issues/6686 to ask for further comments.

**Pull request merged**